sklearn/cluster/optics_.py

```diff
@@ -817,3 +997,397 @@ def _get_leaves(node, arr):
+    """
+
+    # all indices are inclusive (specially at the end)
+    n_samples = len(reachability_plot)
+    # add an inf to the end of reachability plot
+    # but process the data only until the last actual point
+    # this is fine since the last point is considered upward anyway
+    reachability_plot = np.array(reachability_plot)
+    reachability_plot = np.hstack((reachability_plot, np.inf))
+
+    if min_cluster_size <= 1:
+        min_cluster_size = max(2, min_cluster_size * n_samples)
+
+    ixi = 1 - xi
+    sdas = list()
+    clusters = list()
+    index = int(0)
+    mib = 0.  # maximum in between
+    while index + 1 < n_samples:
+        # print("index", index)
+        # print("r", reachability_plot[index])
+        mib = max(mib, reachability_plot[index])
+        # print("mib up there:", mib)
+
+        # check if a steep downward area starts
+        if _steep_downward(reachability_plot, index, ixi):
+            # print("steep downward")
+            # print("sdas", sdas)
+            # print("filter mib:", mib)
+            sdas = _update_fileter_sdas(sdas, mib, ixi)
+            # print("sdas", sdas)
+            D_start = index
+            index, end = _extend_downward(reachability_plot, D_start, ixi,
+                                          min_samples, n_samples)
+            D = _Area(start=D_start, end=end,
+                      maximum=reachability_plot[D_start], mib=0.)
+            # print("D", D, "r.s %.4g" % reachability_plot[D.start],
+            #       "r.e %.4g" % reachability_plot[D.end])
+            sdas.append(D)
+            mib = reachability_plot[index]
+
+        elif _steep_upward(reachability_plot, index, ixi):
+            # print("steep upward")
+            # print("sdas", sdas)
+            # print("filter mib:", mib)
+            sdas = _update_fileter_sdas(sdas, mib, ixi)
+            # print("sdas", sdas)
+            U_start = index
+            index, end = _extend_upward(reachability_plot, U_start, ixi,
+                                        min_samples, n_samples)
+            U = _Area(start=U_start, end=end, maximum=reachability_plot[end],
+                      mib=-1)
+            # if np.isinf(reachability_plot[index + 1]):
+            #     U.maximum = np.inf
+            #     index += 1
+            # print("U", U, "r.s %.4g" % reachability_plot[U.start],
+            #       "r.e %.4g" % reachability_plot[U.end])
+            mib = reachability_plot[end - 1]
+            # print('mib %.4g' % mib)
+            # print(sdas)
+
+            U_clusters = list()
+            for D in sdas:
+                c_start = D.start
+                c_end = min(U.end, n_samples - 1)
+                # print("D", D, "U", U)
+                # print("start", c_start, "end", c_end)
+
+                # 3.b
+                if D.mib > mib * ixi:
+                    continue
+                # print("3b pass")
+
+                # 4
+                if D.maximum * ixi >= U.maximum:
```

Comment:
I'm a bit confused about the use of U.maximum here. Is `U.maximum == reachability_plot[U.end]`? Or `reachability_plot[U.end + 1]` which seems to be what is asked for in the paper?